npm/pkg/dataplane/dataplane.go

```diff
@@ -135,19 +183,126 @@ func (dp *DataPlane) AddPolicy(policies *policies.NPMNetworkPolicy) error {
+	if !ok {
+		klog.Infof("[DataPlane] Policy %s is not found. Might been deleted already", policyName)
+		return nil
+	}
+	// Use the endpoint list saved in cache for this network policy to remove
+	err := dp.policyMgr.RemovePolicy(policy.Name, nil)
 	if err != nil {
 		return fmt.Errorf("[DataPlane] error while removing policy: %w", err)
 	}
+	// Remove references for Rule IPSets first
+	err = dp.deleteIPSetReferences(policy.RuleIPSets, policy.Name, ipsets.NetPolType)
+	if err != nil {
+		return err
+	}
+
+	// Remove references for Selector IPSets
+	err = dp.deleteIPSetReferences(policy.PodSelectorIPSets, policy.Name, ipsets.SelectorType)
+	if err != nil {
+		return err
+	}
+
+	err = dp.ApplyDataPlane()
+	if err != nil {
+		return fmt.Errorf("[DataPlane] error while applying dataplane: %w", err)
+	}
+
 	return nil
 }
 
 // UpdatePolicy takes in updated policy object, calculates the delta and applies changes
 // onto dataplane accordingly
-func (dp *DataPlane) UpdatePolicy(policies *policies.NPMNetworkPolicy) error {
-	err := dp.policyMgr.UpdatePolicy(policies)
+func (dp *DataPlane) UpdatePolicy(policy *policies.NPMNetworkPolicy) error {
+	klog.Infof("[DataPlane] Update Policy called for %s", policy.Name)
+	ok := dp.policyMgr.PolicyExists(policy.Name)
+	if !ok {
+		klog.Infof("[DataPlane] Policy %s is not found. Might been deleted already", policy.Name)
+		return dp.AddPolicy(policy)
+	}
+
+	// TODO it would be ideal to calculate a diff of policies
+	// and remove/apply only the delta of IPSets and policies
+
+	// Taking the easy route here, delete existing policy
+	err := dp.policyMgr.RemovePolicy(policy.Name, nil)
+	if err != nil {
+		return fmt.Errorf("[DataPlane] error while updating policy: %w", err)
+	}
+	// and add the new updated policy
+	err = dp.policyMgr.AddPolicy(policy, nil)
 	if err != nil {
 		return fmt.Errorf("[DataPlane] error while updating policy: %w", err)
 	}
 	return nil
 }
+
+func (dp *DataPlane) addIPSetReferences(sets map[string]*ipsets.IPSet, netpolName string, referenceType ipsets.ReferenceType) error {
+	// Create IPSets first along with reference updates
+	for _, set := range sets {
+		dp.ipsetMgr.CreateIPSet(set.Name, set.Type)
+		err := dp.ipsetMgr.AddReference(set.Name, netpolName, referenceType)
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO is there a possibility for a list set of selector referencing rule ipset?
+	// if so this below addition would throw an error because rule ipsets are not created
+	// Check if any list sets are provided with members to add
+	for _, set := range sets {
+		// Check if any 2nd level IPSets are generated by Controller with members
+		// Apply members to the list set
+		if set.Kind == ipsets.ListSet {
+			if len(set.MemberIPSets) > 0 {
+				memberList := []string{}
+				for _, memberSet := range set.MemberIPSets {
+					memberList = append(memberList, memberSet.Name)
+				}
+				err := dp.ipsetMgr.AddToList(set.Name, memberList)
```

Comment:
I'm confused to the purpose of this code? Is this code adding the members of a list to the list itself?

Reply:
We don't allow the controller's to modify the IPSet objects themselves so I don't think we need this loop

Reply:
IPSets in NPMnetworkPolicy object are what is TRANSLATED from the policy and not what is the actual state of an ipset in cache or dataplane. In Multivalue scenarios translate policy will return a NESTED IPSet with some labels as members, in that we will have to create this list and add relevant members to this new list.